src/date.c

```
@@ -1,18 +1,1006 @@
+  int month; /* 0-11 */
+  int day; /* 1-31 */
+  int hour; /* 0-23 */
+  int min; /* 0-59 */
+  int sec; /* 0-59 */
+  int msec; 
+  int dayofweek; /* 0-6 */
+};
+
+static int d_istimeinvalid(etime_t* time) {
+  return isnan(*time);
+}
+
+/*+++ this functions is used to get current date/time & timezone */
+
+static void d_gettime(etime_t* time) {
+  struct timeval tv;
+  gettimeofday(&tv, NULL);
+  *time = (etime_t)tv.tv_sec * 1000 + (etime_t)tv.tv_usec / 1000;
+}
+
+static int d_gettimezone() {
+  return (int)timezone/60; /* TODO(alashkin): check non-OSX */
+}
+
+static char* d_gettzname() {
+  return tzname[0]; /* TODO(alashkin): check non-OSX */
+}
+
+static etime_t d_mktime_impl(struct timeparts* tp) {
+  return ecma_MakeDate(ecma_MakeDay(tp->year, tp->month, tp->day),
+                       ecma_MakeTime(tp->hour, tp->min, tp->sec, tp->msec));
+}
+
+/*++++ libc mktime alternatives ++++*/
+
+static etime_t d_lmktime(struct timeparts* tp) {
+  return ecma_UTC(d_mktime_impl(tp));
+}
+
+static etime_t d_gmktime(struct timeparts* tp) {
+  return d_mktime_impl(tp);
+}
+
+typedef etime_t (*fmaketime)(struct timeparts* );
+
+/*++++ libc gmtime & localtime alternatives ++++*/
+static void d_gmtime(etime_t* t, struct timeparts* tp) {
+  tp->year = ecma_YearFromTime_s(*t);
+  tp->month = ecma_MonthFromTime(*t, tp->year);
+  tp->day = ecma_DateFromTime(*t, tp->year);
+  tp->hour = ecma_HourFromTime(*t);
+  tp->min = ecma_MinFromTime(*t);
+  tp->sec = ecma_SecFromTime(*t);
+  tp->msec = ecma_msFromTime(*t);
+  tp->dayofweek = ecma_WeekDay(*t);
+}
+
+static void d_localtime(etime_t* time, struct timeparts* tp) {
+  etime_t local_time = ecma_LocalTime(*time);
+  d_gmtime(&local_time, tp);
+}
+
+typedef void (*fbreaktime)(etime_t* , struct timeparts* );
+
+/*++++ some isXXX helpers ++++*/
+
+static int d_isnumberNAN(struct v7 *v7, val_t obj) {
+  return (i_value_of(v7, obj) == V7_TAG_NAN);
+}
+
+static void d_isobjvalidforstirng(struct v7* v7, val_t obj) {
```

I'd prefer `check...` instead of `is...` for functions that throw exceptions instead of returning bool.